### PR TITLE
[JENKINS-23271] - Prevent early deallocation of the Proc instance by GC in ProcStarter#join()

### PR DESCRIPTION
**Pull request by oleg-nenashev**
Just realized I have not proposed a hackish fix for [JENKINS-23271](https://issues.jenkins-ci.org/browse/JENKINS-23271) yet. It was implemented a while ago, but seems to be still actual. Have not tested it with the latest core BTW.

It is a hackish way, which likely prevents a preliminary deallocation of the spawned `RemoteProc` instance, which we see in JENKINS-23271. Proc instance was not actually required in the original code since we were creating and using `RemoteInvocationHandler` wrapper only, and the theory discussed with @stephenc was that object gets removed by Java8 garbage collector before we get into join().

This fix enforces the persistency of `ProcStarter#start()` result by adding logging and the enforced volatile field (maybe the last one is not really required, but JIT compiler in Java implementations may be smart enough to skip unused loggers)

This is a pretty old fix from August, which has been soak tested on my instance for several weeks (mid-August => Jenkins World). On the reference instance (just a small Jenkins 2.7.1 instance with 4 agents and very frequent builds with `CommandInterpreter steps`) I saw 2 failures over the period. On the fixed instance - 0. It does not proof anything, but at least the fix was soak tested a bit.

@reviewbybees @stephenc

### CODE REVIEW

File: core/src/main/java/hudson/Launcher.java [hunk 1]
```diff
@@ -165,6 +167,14 @@ public Computer getComputer() {
         protected boolean reverseStdin, reverseStdout, reverseStderr;
 
         /**
+         * Just an instance holder for {@link #join()}, which creates a temporary {@link Proc} instance.
+         * Usage of {@code volatile} should prevent preliminary deallocation of the process object during the {@link Proc#join()} call.
+         */
+        @CheckForNull
+        @GuardedBy("this")
+        private volatile Proc procHolderForJoin;
```
Comment (stephenc) [inline core/src/main/java/hudson/Launcher.java]: I think this is unnecessary for the fix

File: core/src/main/java/hudson/Launcher.java [hunk 2]
```diff
@@ -383,9 +393,24 @@ public Proc start() throws IOException {
...
         /**
          * Starts the process and waits for its completion.
+         * This method stores internal context and hence it is synchronized
+         * @return Return code of the invoked process
+         * @throws IOException Operation error (e.g. remote call failure)
+         * @throws InterruptedException The process has been interrupted
          */
-        public int join() throws IOException, InterruptedException {
-            return start().join();
+        public synchronized int join() throws IOException, InterruptedException {
```
Comment (stephenc) [inline core/src/main/java/hudson/Launcher.java]: I think what we really want to do is:

```
Proc proc = start();
try {
  return proc.join();
} finally {
  if (proc.isAlive()) { // will never happen but this forces Proc to not be removed and early GC by escape analysis
    LOGGER.log(Level.WARNING, "Process not finished after call to join() completed");
  }
}
```

Which has the advantage of reducing the memory footprint and localizing the hack to just a single method

File: core/src/main/java/hudson/Launcher.java [hunk 3]
```diff
@@ -383,9 +393,24 @@ public Proc start() throws IOException {
 
         /**
          * Starts the process and waits for its completion.
+         * This method stores internal context and hence it is synchronized
```
Comment (stephenc) [inline core/src/main/java/hudson/Launcher.java]: Adding a use case for the objects public monitor is never a good plan IMHO

**Review by stephenc — COMMENTED** [review]
I think the fix you have suggested is the wrong fix, but given that you have soak tested, if you want to take the pragmatic route and push your change with a follow-up to apply my suggestion after further soak testing (assuming my suggestion works) and you add the TODO comments to indicate such a plan... then I am fine

**Comment by oleg-nenashev** [comment]
@stephenc I agree with the comments. As you may have seen in the summary, I also do not see much need in the volatile field. It was just a hack I had in soak testing. I will remove the field

**Comment by oleg-nenashev** [comment]
CC @aheritier, who was pinging me about this issue for months :(

**Comment by aheritier** [comment]
Yeah @oleg-nenashev !!!
LGTM 🐝

**Comment by daniel-beck** [comment]
Preliminary? 

> denoting an action or event preceding or done in preparation for something fuller or more important

**Comment by oleg-nenashev** [comment]
Fixed my Inglesh

**Review by stephenc — CHANGES_REQUESTED** [review]
The `assert` is dodgy IMHO. I'd prefer the `finally` pattern

File: core/src/main/java/hudson/Launcher.java [hunk 4]
```diff
@@ -393,24 +385,23 @@ public Proc start() throws IOException {
...
-            } finally {
-                procHolderForJoin = null;
+            final Proc procHolderForJoin = start();
+            LOGGER.log(Level.FINER, "Started the process {0}", procHolderForJoin);
+            final int returnCode = procHolderForJoin.join();
+            if (LOGGER.isLoggable(Level.FINER)) {
+                LOGGER.log(Level.FINER, "Process {0} has finished with the return code {1}", new Object[] {procHolderForJoin, returnCode});
+            }
+            if (procHolderForJoin.isAlive()) { // Should never happen but this forces Proc to not be removed and early GC by escape analysis
+                assert false : "Process not finished after call to join() completed";
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: Finally block is also not a silver bullet, right?

Comment (stephenc) [inline core/src/main/java/hudson/Launcher.java]: So the contract of finally means that the JVM cannot evaluate it out of order, and then by not bothering with an assert, rather running to warning log level then escape analysis cannot remove the if... plus while it should never happen, we want to know if it does... an assertion error is not a good way to have that happen...

Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: OK. Let's do finally with assert :)

File: core/src/main/java/hudson/Launcher.java [hunk 5]
```diff
@@ -383,9 +385,26 @@ public Proc start() throws IOException {
...
+            LOGGER.log(Level.FINER, "Started the process {0}", procHolderForJoin);
+            try {
+                final int returnCode = procHolderForJoin.join();
+                if (LOGGER.isLoggable(Level.FINER)) {
+                    LOGGER.log(Level.FINER, "Process {0} has finished with the return code {1}", new Object[]{procHolderForJoin, returnCode});
+                }
+                return returnCode;
+            } finally {
+                if (procHolderForJoin.isAlive()) { // Should never happen but this forces Proc to not be removed and early GC by escape analysis
+                    assert false : "Process not finished after call to join() completed";
```
Comment (stephenc) [inline core/src/main/java/hudson/Launcher.java]: And still you leave the exploding bomb of an `assert false`

**Review by stephenc — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/Launcher.java (hunk 5 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Launcher.java]: it is bad? I will remove it just to unblock anyway

**Review by daniel-beck — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
I assume the last comment from @stephenc is addressed. I will merge it tomorrow if the tests pass

**Review by stephenc — APPROVED** [review]
:bee:

**Comment by oleg-nenashev** [comment]
CI passed, merging

**Pull request merged**